### PR TITLE
V3 update

### PR DESCRIPTION
**Pull request by Raistlfiren**
Here is a push of the Google Analytics plugin for Bolt V3.

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -1,24 +1,36 @@
 {
     "name": "bolt/googleanalytics",
-    "description": "",
+    "description": "A Google Analytics Extension for Bolt",
     "type": "bolt-extension",
     "keywords": [],
     "require": {
-        "bolt/bolt": ">=2.0.0,<3.0.0"
+        "bolt/bolt": "^3.0.0"
```
Comment (GawainLynch) [inline composer.json]: Only needs `"bolt/bolt": "^3.0"`

File: src/GoogleAnalyticsExtension.php [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+    protected function registerServices(Application $app)
+    {
+        //Register translations in the service provider.
+        $translationDirectory = $this->getBaseDirectory()->getDir('translations');
+        if ($translationDirectory->exists()) {
+            foreach ($translationDirectory->getContents(true) as $fileInfo) {
+                if ($fileInfo->isFile()) {
+                    list($domain, $extension) = explode('.', $fileInfo->getFilename());
+
+                    $path = $app['resources']->getPath('extensions' . DIRECTORY_SEPARATOR . $fileInfo->getPath());
```
Comment (GawainLynch) [inline src/GoogleAnalyticsExtension.php]: For reference, you can just `$path = $app['resources']->getPath('extensions/'  . $fileInfo->getPath());` as the path will be normalised in the resource manager.

File: src/GoogleAnalyticsExtension.php [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+
+        parent::registerServices($app);
+    }
+
+    protected function registerBackendRoutes(ControllerCollection $collection)
+    {
+        $app = $this->getContainer();
+
+        //Block unauthorized access...
+        if ($app['users']->isAllowed('dashboard')) {
```
Comment (GawainLynch) [inline src/GoogleAnalyticsExtension.php]: This should be done in the controller/route

File: src/GoogleAnalyticsExtension.php [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+            $collection->match('/extensions/google-analytics', [$this, 'googleAnalytics']);
+        }
+    }
+
+    protected function registerTwigPaths()
+    {
+        return [
+            'templates'
+        ];
+    }
```
Comment (GawainLynch) [inline src/GoogleAnalyticsExtension.php]: Not needed as the trait adds it automatically for you

File: src/GoogleAnalyticsExtension.php [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+            $analyticsCode
+        ];
+    }
+
+    public function googleAnalytics(Application $app, Request $request)
+    {
+        $config = $this->getConfig();
+
+        $data = [
+            "locale" => substr($this->app['locale'], 0, 2),
```
Comment (GawainLynch) [inline src/GoogleAnalyticsExtension.php]: `$this->app` isn't valid here :wink:

File: src/GoogleAnalyticsExtension.php [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+        $data = [
+            'webproperty' => $config['webproperty']
+        ];
+
+        if ($config['universal']) {
+            $data['domainname'] = $config['universal_domainname'];
+            return $app['twig']->render("universal.twig", $data);
+        }
+
+        $data['domainname'] = $_SERVER['HTTP_HOST'];
```
Comment (GawainLynch) [inline src/GoogleAnalyticsExtension.php]: Say "NO!" to super globals…

```
        $data['domainname'] = $app['request_stack']->getCurrentRequest()->server->get('HTTP_HOST');
```

File: src/GoogleAnalyticsExtension.php [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+        if (empty($config['ga_profile_id'])) {
+            return "ga_profile_id not set in config.yml.";
+        }
+
+        $service_account_email = $config['service_account_email']; //Email Address
+        $key_file_location = $config['key_file_location']; //key.p12
+
+        $path = $app['resources']->getPath('config');
+
+        $path .= DIRECTORY_SEPARATOR . 'extensions' . DIRECTORY_SEPARATOR . $key_file_location;
```
Comment (GawainLynch) [inline src/GoogleAnalyticsExtension.php]:
```
$path = $app['resources']->getPath('extensionsconfig/' . $key_file_location);
```

File: src/GoogleAnalyticsExtension.php [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+        if (!file_exists($path)) {
+            return "Key file not found in app/config/extenstions/!";
+        }
+
+        require_once(__DIR__.'/../Google/autoload.php');
+
+        // Create and configure a new client object.
+        $client = new \Google_Client();
+        $client->setApplicationName("HelloAnalytics");
+        $analytics = new \Google_Service_Analytics($client);
```
Comment (GawainLynch) [inline src/GoogleAnalyticsExtension.php]: `$analytics` doesn't get used… if WiP ignore me :smile:

**Comment by GawainLynch** [comment]
Overall great work… Thank you!

**Comment by Raistlfiren** [comment]
Thanks as always @GawainLynch! I have updated my code as you requested. It seems as though you need to use `$this->renderTemplate(...);` in order to use templating. Whoops, I didn't know! :)

**Comment by GawainLynch** [comment]
Legendary @Raistlfiren :tada:

**Comment by bobdenotter** [comment]
Ok, I _finally_ figured out why this repository wasn't working for me. My *adblocker* was blocking ajaxy stuff because of the 'GoogleAnalytics' in the URL. :-/

**Comment by Raistlfiren** [comment]
Wow, you are right! Nice job on catching that! How did you figure that out?

**Comment by bobdenotter** [comment]
@Raistlfiren Just poking around, until it suddenly clicked.. 

I've been trying the update, but I couldn't get it to work just yet.. Any idea what's happening here? 

screen shot 2016-06-21 at 20 10 57


I'm going to push some typo's i've encountered, though.

**Comment by Raistlfiren** [comment]
@bobdenotter bump

**Comment by bobdenotter** [comment]
@Raistlfiren Sorry mate. Last time i've tried it, i've not been able to get this up and running.. :-/

I'm not sure yet if i'm dumb, if Google's stuff is overcomplicated, or if it needs more documentation. (or a combination of these things)

**Comment by bobdenotter** [comment]
Merging in! We'll do some more minor cleanup, and then we can release it.. Awesome work, @Raistlfiren ! 👍

**Pull request merged**